include/deal.II/lac/utilities.h

```diff
@@ -147,6 +168,73 @@ namespace Utilities
       return diagonal[k-1] + f->l2_norm();
     }
 
+
+    template <typename OperatorType, typename VectorType>
+    void Chebyshev_filter(VectorType &x,
+                          const OperatorType &op,
+                          const unsigned int degree,
+                          const double a,
+                          const double b,
+                          const double a_L,
+                          VectorMemory<VectorType> &vector_memory)
+    {
+      Assert (degree > 0,
+              ExcNotImplemented());
+
+      const bool scale = !numbers::is_nan(a_L);
+      Assert (a < b,
+              ExcMessage("Lower bound of the unwanted spectrum should be smaller than the highest bound."));
+
+      Assert (a_L <= a || a_L >= b || !scale,
+              ExcMessage("Scaling point should be outside of the filtering range."));
+
+      // Setup auxiliary vectors:
+      typename VectorMemory<VectorType>::Pointer p_y(vector_memory);
+      typename VectorMemory<VectorType>::Pointer p_yn(vector_memory);
+
+      p_y->reinit(x);
+      p_yn->reinit(x);
+
+      // convinience to avoid pointers
+      VectorType &y   = *p_y;
+      VectorType &yn  = *p_yn;
+
+      // Below is an implementation of
+      // Algorithm 3.2 in Zhou et al, Journal of Computational Physics 274 (2014) 770-782
+      // with **a bugfix for sigma1**. Here is the original algorithm verbatim:
+      //
+      // [Y]=Chebyshev_filter_scaled(X, m, a, b, aL).
+      // e=(b−a)/2; c=(a+b)/2; σ=e/(c−aL); τ=2/σ;
+      // Y=(H∗X−c∗X)∗(σ/e);
+      // for i=2 to m do
+      //   σnew =1/(τ −σ);
+      //   Yt =(H∗Y−c∗Y)∗(2∗σnew/e)−(σ∗σnew)∗X;
+      //   X =Y; Y =Yt; σ =σnew;
+
+      const double e = (b-a)/2.;
+      const double c = (a+b)/2.;
+      const double alpha = 1./e;
+      const double beta = - c/e;
+
+      const double sigma1 = e/(a_L - c); // BUGFIX which is relevant for odd degrees
+      double sigma = scale ? sigma1 : 1.;
+      const double tau = 2./sigma;
+      op.vmult(y,x);
+      y.sadd(alpha*sigma, beta*sigma, x);
+
+      for (unsigned int i = 2; i <= degree; ++i)
+        {
+          const double sigma_new = scale ? 1./(tau-sigma) : 1.;
+          op.vmult(yn,y);
+          yn.sadd(2.*alpha*sigma_new, 2.*beta*sigma_new, y, -sigma*sigma_new, x);
+          x.swap(y);
+          y.swap(yn);
```

Comment:
that works. do we have move operators for all vector classes? If so, we could write
```
  x = std::move(y);
```
but I suspect that we can't move PETSc or Trilinos objects, for example, so leaving it as is is probably the right thing to do.